drivers/sensor/bosch/bmm350/bmm350.c

```diff
@@ -0,0 +1,984 @@
+	int8_t ret = 0;
+	uint8_t rx_buf[3] = {0x00};
+
+	if (ret == BMM350_OK) {
+		ret = bmm350_reg_read(dev, BMM350_REG_PMU_CMD, &rx_buf[0], 3);
+		if (ret == BMM350_OK) {
+			if (rx_buf[2] > BMM350_PMU_CMD_NM_TC) {
+				ret = BMM350_E_INVALID_CONFIG;
+			}
+
+			if ((ret == BMM350_OK) && ((rx_buf[2] == BMM350_PMU_CMD_NM) ||
+						   (rx_buf[2] == BMM350_PMU_CMD_UPD_OAE))) {
+				/* Set PMU command configuration */
+				ret = bmm350_reg_write(dev, BMM350_REG_PMU_CMD, BMM350_PMU_CMD_SUS);
+			}
+
+			if (ret == BMM350_OK) {
+				ret = set_powermode(powermode, dev);
+			}
+		}
+	}
+
+	return ret;
+}
+
+/*!
+ * used to perform the magnetic reset of the sensor
+ * which is necessary after a field shock ( 400mT field applied to sensor )
+ */
+static int8_t bmm350_magnetic_reset_and_wait(const struct device *dev)
+{
+	/* Variable to store the function result */
+	int8_t ret = 0;
+	struct bmm350_pmu_cmd_status_0 pmu_cmd_stat_0 = {0};
+	uint8_t restore_normal = BMM350_DISABLE;
+
+	/* Read PMU CMD status */
+	ret = bmm350_get_pmu_cmd_status_0(&pmu_cmd_stat_0, dev);
+	LOG_DBG("get status result 0:%d\n", ret);
+
+	/* Check the powermode is normal before performing magnetic reset */
+	if ((ret == BMM350_OK) && (pmu_cmd_stat_0.pwr_mode_is_normal == BMM350_ENABLE)) {
+		restore_normal = BMM350_ENABLE;
+		/* Reset can only be triggered in suspend */
+		ret = bmm350_set_powermode(BMM350_SUSPEND_MODE, dev);
+		LOG_DBG("set power mode 0:%d\n", ret);
+	}
+	if (ret == BMM350_OK) {
+		/* Set BR to PMU_CMD register */
+		ret = bmm350_reg_write(dev, BMM350_REG_PMU_CMD, BMM350_PMU_CMD_BR);
+		k_usleep(BMM350_BR_DELAY);
+	}
+
+	if (ret == BMM350_OK) {
+		/* Verify if PMU_CMD_STATUS_0 register has BR set */
+		ret = bmm350_get_pmu_cmd_status_0(&pmu_cmd_stat_0, dev);
+		LOG_DBG("get status result 1:%d\n", ret);
+		if ((ret == BMM350_OK) &&
+		    (pmu_cmd_stat_0.pmu_cmd_value != BMM350_PMU_CMD_STATUS_0_BR)) {
+			ret = BMM350_E_PMU_CMD_VALUE;
+		}
+	}
+
+	if (ret == BMM350_OK) {
+		/* Set FGR to PMU_CMD register */
+		ret = bmm350_reg_write(dev, BMM350_REG_PMU_CMD, BMM350_PMU_CMD_FGR);
+		k_usleep(BMM350_FGR_DELAY);
+	}
+	if (ret == BMM350_OK) {
+		/* Verify if PMU_CMD_STATUS_0 register has FGR set */
+		ret = bmm350_get_pmu_cmd_status_0(&pmu_cmd_stat_0, dev);
+		LOG_DBG("get status result 2:%d\n", ret);
+
+		if ((ret == BMM350_OK) &&
+		    (pmu_cmd_stat_0.pmu_cmd_value != BMM350_PMU_CMD_STATUS_0_FGR)) {
+			ret = BMM350_E_PMU_CMD_VALUE;
+		}
+	}
+
+	if ((ret == BMM350_OK) && (restore_normal == BMM350_ENABLE)) {
+		ret = bmm350_set_powermode(BMM350_NORMAL_MODE, dev);
+		LOG_DBG("set power mode 1:%d\n", ret);
+	} else {
+		if (ret == BMM350_OK) {
+			/* Reset PMU_CMD register */
+			ret = bmm350_reg_write(dev, BMM350_REG_PMU_CMD, 0x00);
+		}
+	}
+	return ret;
+}
+/*!
+ * @brief This API is used to read uncompensated mag and temperature data.
+ */
+int8_t bmm350_read_uncomp_mag_temp_data(struct bmm350_raw_mag_data *raw_data,
+					const struct device *dev)
```

Comment:
nit: usually, I always see the `const struct device *` as the first arg for every function. Just something that seems to not follow a pattern